java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -0,0 +1,261 @@
+      constraintManager = context.getConstraintManager();
+    }
+
+    private void addZeroConstraint(boolean isZero) {
+      constraintManager.setValueFactory((id, node) -> new ZeroSymbolicValue(id, isZero));
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      boolean varIsZero = isZero(programState, symbolicValues.get(0));
+      boolean exprIsZero = isZero(programState, symbolicValues.get(1));
+
+      switch (tree.kind()) {
+        case MULTIPLY_ASSIGNMENT:
+          addZeroConstraint(exprIsZero);
+          break;
+        case MINUS_ASSIGNMENT:
+        case PLUS_ASSIGNMENT:
+        case DIVIDE_ASSIGNMENT:
+        case REMAINDER_ASSIGNMENT:
+          addZeroConstraint(varIsZero);
+          break;
+        default:
+          // do nothing
+          break;
+      }
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      if (tree.is(
+        Tree.Kind.MULTIPLY, Tree.Kind.PLUS, Tree.Kind.MINUS, Tree.Kind.DIVIDE, Tree.Kind.REMAINDER,
+        Tree.Kind.GREATER_THAN, Tree.Kind.GREATER_THAN_OR_EQUAL_TO, Tree.Kind.LESS_THAN, Tree.Kind.LESS_THAN_OR_EQUAL_TO)) {
+        List<SymbolicValue> symbolicValues = programState.peekValues(2);
+        SymbolicValue rightOp = symbolicValues.get(0);
+        boolean leftOpIsZero = isZero(programState, rightOp);
+        SymbolicValue leftOp = symbolicValues.get(1);
+        boolean rightOpIsZero = isZero(programState, leftOp);
+        switch (tree.kind()) {
+          case MULTIPLY:
+            addZeroConstraint(leftOpIsZero || rightOpIsZero);
+            break;
+          case PLUS:
+          case MINUS:
+            addZeroConstraint(leftOpIsZero && rightOpIsZero);
+            break;
+          case DIVIDE:
+          case REMAINDER:
+            addZeroConstraint(leftOpIsZero);
+            break;
+          case GREATER_THAN:
+          case GREATER_THAN_OR_EQUAL_TO:
+          case LESS_THAN:
+          case LESS_THAN_OR_EQUAL_TO:
+            programState = programState.addConstraint(leftOp, ZeroConstraint.nonZero(tree.leftOperand()));
+            programState = programState.addConstraint(rightOp, ZeroConstraint.nonZero(tree.rightOperand()));
+            break;
+          default:
+            // do nothing
+            break;
+        }
+      }
+    }
+
+    @Override
+    public void visitUnaryExpression(UnaryExpressionTree tree) {
+      if (tree.is(Tree.Kind.UNARY_MINUS, Tree.Kind.UNARY_PLUS)) {
+        addZeroConstraint(isZero(programState, programState.peekValue()));
+      }
+    }
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    PostStatementVisitor visitor = new PostStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private class PostStatementVisitor extends CheckerTreeNodeVisitor {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2694)